lib/SILOptimizer/Analysis/EscapeAnalysis.cpp

```diff
@@ -1858,23 +1883,28 @@ bool EscapeAnalysis::canEscapeToUsePoint(SILValue V, SILNode *UsePoint,
   if (ConGraph->isUsePoint(UsePoint, Node))
     return true;
 
-  assert(isPointer(V) && "should not have a node for a non-pointer");
-
-  // Check if the object "content" can escape to the called function.
-  // This will catch cases where V is a reference and a pointer to a stored
-  // property escapes.
-  // It's also important in case of a pointer assignment, e.g.
-  //    V = V1
-  //    apply(V1)
-  // In this case the apply is only a use-point for V1 and V1's content node.
-  // As V1's content node is the same as V's content node, we also make the
-  // check for the content node.
-  CGNode *ContentNode = ConGraph->getContentNode(Node);
-  if (ContentNode->escapesInsideFunction(false))
-    return true;
+  if (V->getType().hasReferenceSemantics()) {
+    // In case V is the result of getUnderlyingObject, we also have to check
+    // the content node. Example:
+    //
+    //   %a = ref_element %r
+    //   apply %f(%a)
+    //
+    // The reference %r does not actually escape to the function call. But in
+    // case this API is called like
+    //
+    //   canEscapeToUsePoint(getUnderlyingObject(applyArgument))
+    //
+    // it would yield false, although the projected object address is passed to
+    // the function.
```

Comment:
I don't see how this is correct in the opposite scenario: a pointee is passed to the pointer's use point. To handle that, either the use points need to be forward propagated through pointsTo edges, or this function needs to follow all the pointsTo predecessors (which seems expensive).

Reply:
Use points are forward propagated through all edge types. Not sure if this answers your question.